opensearch-operator/pkg/reconcilers/ismpolicy.go

```diff
@@ -0,0 +1,531 @@
+		reason = "waiting for opensearch cluster to exist"
+		r.recorder.Event(r.instance, "Normal", opensearchPending, reason)
+		retResult = ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: 10 * time.Second,
+		}
+		return
+	}
+	// Check cluster ref has not changed
+	if r.instance.Status.ManagedCluster != nil {
+		if *r.instance.Status.ManagedCluster != r.cluster.UID {
+			reason = "cannot change the cluster a role refers to"
+			retErr = fmt.Errorf("%s", reason)
+			r.recorder.Event(r.instance, "Warning", opensearchRefMismatch, reason)
+			return
+		}
+	}
+	// Check cluster is ready
+	if r.cluster.Status.Phase != opsterv1.PhaseRunning {
+		r.logger.Info("opensearch cluster is not running, requeueing")
+		reason = "waiting for opensearch cluster status to be running"
+		r.recorder.Event(r.instance, "Normal", opensearchPending, reason)
+		retResult = ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: 10 * time.Second,
+		}
+		return
+	}
+
+	r.osClient, err = util.CreateClientForCluster(r.ctx, r.Client, r.cluster, r.osClientTransport)
+	if err != nil {
+		reason := "error creating opensearch client"
+		r.recorder.Event(r.instance, "Warning", opensearchError, reason)
+	}
+	// If PolicyID not provided explicitly, use metadata.name by default
+	if r.instance.Spec.PolicyID == "" {
+		r.instance.Spec.PolicyID = r.instance.Name
+	}
+	// Check ism policy state to make sure we don't touch preexisting ism policy
+	if r.instance.Status.ExistingISMPolicy == nil {
+		var exists bool
+		exists, retErr = services.PolicyExists(r.ctx, r.osClient, r.instance.Spec.PolicyID)
+		if retErr != nil {
+			reason = "failed to get policy status from Opensearch API"
+			r.logger.Error(retErr, reason)
+			r.recorder.Event(r.instance, "Warning", opensearchAPIError, reason)
+			return
+		}
+		if pointer.BoolDeref(r.updateStatus, true) {
+			retErr = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				if err := r.Get(r.ctx, client.ObjectKeyFromObject(r.instance), r.instance); err != nil {
+					return err
+				}
+				r.instance.Status.ExistingISMPolicy = &exists
+				return r.Status().Update(r.ctx, r.instance)
+			})
+			if retErr != nil {
+				reason = fmt.Sprintf("failed to update status: %s", retErr)
+				r.recorder.Event(r.instance, "Warning", statusError, reason)
+				return
+			}
+		} else {
+			// Emit an event for unit testing assertion
+			r.recorder.Event(r.instance, "Normal", "UnitTest", fmt.Sprintf("exists is %t", exists))
+			return
+		}
+	}
+
+	// If ism policy is existing do nothing
+	if *r.instance.Status.ExistingISMPolicy {
+		reason = ismPolicyExists
+		return
+	}
+	ismpolicy, retErr := r.CreateISMPolicyRequest()
+	if retErr != nil {
+		reason = "failed to get create the ism policy request"
+		r.logger.Error(retErr, reason)
+		r.recorder.Event(r.instance, "Warning", opensearchAPIError, reason)
+		return
+	}
+
+	ismResponse, retErr := services.GetPolicy(r.ctx, r.osClient, r.instance.Spec.PolicyID)
```

Comment:
`ismResponse` is not a good name, should be more descriptive, like `existingPolicy`.